lms/static/js/learner_dashboard/models/course_card_model.js

```diff
@@ -1,273 +1,261 @@
-                    var selectedCourseRun = _.findWhere(this.get('course_runs'), {key: courseRunKey});
-                    if (selectedCourseRun) {
-                        // Update the current context to set the course run to the enrolled state
-                        _.each(this.context.course_runs, function(run) {
-                            if (run.key === selectedCourseRun.key) run.is_enrolled = true; // eslint-disable-line no-param-reassign, max-len
-                        });
-                        this.setActiveCourseRun(selectedCourseRun);
-                    }
-                },
-
-                // update marketing url for deep linking if is_mobile_only true
-                updateMarketingUrl: function(courseRun) {
-                    if (courseRun.is_mobile_only === true) {
-                        var marketingUrl = courseRun.marketing_url, // eslint-disable-line vars-on-top
-                            href = marketingUrl,
-                            path,
-                            start;
-
-                        if (marketingUrl.indexOf('course_info?path_id') < 0) {
-                            start = marketingUrl.indexOf('course/');
-
-                            if (start > -1) {
-                                path = marketingUrl.substr(start);
-                            }
-
-                            href = 'edxapp://course_info?path_id=' + path;
-                        }
-
-                        return href;
-                    } else {
-                        return courseRun.marketing_url;
-                    }
-                }
-            });
-        });
-}).call(this, define || RequireJS.define);
+class CourseCardModel extends Backbone.Model {
+  initialize(data) {
+    if (data) {
+      this.context = data;
+      this.setActiveCourseRun(this.getCourseRun(data), data.user_preferences);
+    }
+  }
+
+  getCourseRun(course) {
+    const enrolledCourseRun = _.findWhere(course.course_runs, { is_enrolled: true });
+    const openEnrollmentCourseRuns = this.getEnrollableCourseRuns();
+    let desiredCourseRun;
+
+    // If the learner has an existing, unexpired enrollment,
+    // use it to populate the model.
+    if (enrolledCourseRun && !course.expired) {
+      desiredCourseRun = enrolledCourseRun;
+    } else if (openEnrollmentCourseRuns.length > 0) {
+      if (openEnrollmentCourseRuns.length === 1) {
+        desiredCourseRun = openEnrollmentCourseRuns[0];
+      } else {
+        desiredCourseRun = CourseCardModel.getUnselectedCourseRun(openEnrollmentCourseRuns);
+      }
+    } else {
+      desiredCourseRun = CourseCardModel.getUnselectedCourseRun(course.course_runs);
+    }
+
+    return desiredCourseRun;
+  }
+
+  isEnrolledInSession() {
+    // Returns true if the user is currently enrolled in a session of the course
+    return _.findWhere(this.context.course_runs, { is_enrolled: true }) !== undefined;
+  }
+
+  static getUnselectedCourseRun(courseRuns) {
```

Comment:
Why does this special butterfly get to be static?

Reply:
It never uses 'this', so eslint complains if it's not static.

Reply:
JK, I went and educated myself.